### PR TITLE
Adjust locations of setting the policy in train/eval mode

### PR DESCRIPTION
**Pull request by maxhuettenrauch**
Addresses #1122:
* We Introduced a new flag `is_within_training_step` which is enabled by the training algorithm when within a training step, where a training step encompasses training data collection and policy updates. This flag is now used by algorithms to decide whether their `deterministic_eval` setting should indeed apply instead of the torch training flag (which was abused!).
* The policy's training/eval mode (which should control torch-level learning only) no longer needs to be set in user code in order to control collector behaviour (this didn't make sense!). The respective calls have been removed.
* The policy should, in fact, always be in evaluation mode when applying data collection, as there is no reason to ever have gradient accumulation enabled for any type of rollout. We thus specifically set the policy to evaluation mode in Collector.collect. Further, it never makes sense to compute gradients during collection, so the possibility to pass `no_grad=False` was removed.

Further changes:
- Base class for collectors: `BaseCollector`
- New util context managers `in_eval_mode` and `in_train_mode` for torch modules. 
- `reset` of `Collectors` now returns `obs` and `info`. 
- `no-grad` no longer accepted as kwarg of `collect`
- Removed deprecations of `0.5.1` (will likely not affect anyone) and the unused `warnings` module.

### CODE REVIEW

**Comment by MischaPanch** [comment]
I'll add the context manager for policies (maybe torch already has one), as discussed, and then this is good to go. Thanks for the extension @maxhuettenrauch !

**Review by MischaPanch — DISMISSED** [review]

**Comment by MischaPanch** [comment]
@ChenDRAG Once finalized (there's still a bit of review and discussion to be done), I'd merge this PR without squashing if you don't mind. It contains somehow unrelated changes and the history would be easier to bisect on commit level if we don't squash

File: tianshou/policy/modelfree/dqn.py [hunk 1]
```diff
@@ -210,6 +210,8 @@ def forward(
         return cast(ModelOutputBatchProtocol, result)
 
     def learn(self, batch: RolloutBatchProtocol, *args: Any, **kwargs: Any) -> TDQNTrainingStats:
+        # set policy in train mode
```
Comment (opcode81) [inline tianshou/policy/modelfree/dqn.py]: I would remove all of these comments, because this is clear from the docstring of `train`.

File: tianshou/policy/modelfree/a2c.py [hunk 2]
```diff
@@ -165,6 +165,9 @@ def learn(  # type: ignore
         *args: Any,
         **kwargs: Any,
     ) -> TA2CTrainingStats:
+        # set policy in train mode
```
Comment (opcode81) [inline tianshou/policy/modelfree/a2c.py:168]: Comment should be replaced with a docstring below the definition.

File: tianshou/policy/base.py [hunk 3]
```diff
@@ -225,6 +225,8 @@ def __init__(
         self.action_scaling = action_scaling
         self.action_bound_method = action_bound_method
         self.lr_scheduler = lr_scheduler
+        # whether the policy is in evaluation mode
```
Comment (opcode81) [inline tianshou/policy/base.py]: Comment should be replaced with a docstring below the definition.

File: tianshou/policy/base.py [hunk 4]
```diff
@@ -225,6 +225,8 @@ def __init__(
...
+        self.is_eval = False  # TODO: remove in favor of kwarg in compute_action/forward?
```
Comment (opcode81) [inline tianshou/policy/base.py]: This breaks compatibility with persisted policies. Must implement `__setstate__`

File: tianshou/data/collector.py [hunk 5]
```diff
@@ -242,18 +236,168 @@ def reset_buffer(self, keep_statistics: bool = False) -> None:
     def reset_env(
         self,
         gym_reset_kwargs: dict[str, Any] | None = None,
-    ) -> None:
+    ) -> tuple[np.ndarray, np.ndarray]:
```
Comment (opcode81) [inline tianshou/data/collector.py:239]: Return value should be documented in docstring.

File: tianshou/data/collector.py [hunk 6]
```diff
@@ -242,18 +236,168 @@ def reset_buffer(self, keep_statistics: bool = False) -> None:
...
 
+    def close(self) -> None:
+        super().close()
+        self._pre_collect_obs_RO = None
+        self._pre_collect_info_R = None
+
+    def reset_env(
+        self,
+        gym_reset_kwargs: dict[str, Any] | None = None,
+    ) -> tuple[np.ndarray, np.ndarray]:
```
Comment (opcode81) [inline tianshou/data/collector.py:385]: Docstring should be removed for overridden method. Several more occurrences below.

File: tianshou/data/collector.py [hunk 7]
```diff
@@ -242,18 +236,168 @@ def reset_buffer(self, keep_statistics: bool = False) -> None:
...
+    #     For n_step, we always use all envs to collect the data, while for n_episode,
+    #     R will be at most n_episode at the beginning, but can decrease during the collection.
+    # O - dimension(s) of observations
+    # A - dimension(s) of actions
+    # H - dimension(s) of hidden state
+    # D - number of envs that reached done in the current collect iteration. Only relevant in n_episode case.
+    # S - number of surplus envs, i.e. envs that are ready but won't be used in the next iteration.
+    #     Only used in n_episode case. Then, R becomes R-S.
+
+    # set the policy's modules to eval mode (this affects modules like dropout and batchnorm) and the policy
```
Comment (MischaPanch) [inline tianshou/data/collector.py]: memo for self - remove this comment (no longer true)

**Comment by opcode81** [comment]
Notebooks and documentation have not been updated and still use the old way of setting train/eval mode.

File: tianshou/data/collector.py [hunk 8]
```diff
@@ -242,18 +236,168 @@ def reset_buffer(self, keep_statistics: bool = False) -> None:
...
+
+        :return: The collected stats
+        """
+        # check that exactly one of n_step or n_episode is set and that the other is larger than 0
+        self._validate_n_step_n_episode(n_episode, n_step)
+
+        if reset_before_collect:
+            self.reset(reset_buffer=False, gym_reset_kwargs=gym_reset_kwargs)
+
+        policy_mode_context = in_eval_mode if eval_mode else in_train_mode
```
Comment (opcode81) [inline tianshou/data/collector.py]: Suggest to add a context manager which accepts a flag, such that this usage pattern is simplified.

Comment (opcode81) [reply tianshou/data/collector.py]: Actually, Max and I have now concluded that there is really no case where the collector should ever use the policy in train mode. The _only_ reason why it currently needs to be in train mode is that some policies have a `deterministic_eval` flag, which is True by default (e.g. SAC) and which should basically never be active when collecting data for training - BUT which is deactivated based on the policy's "training" status only.

Comment (opcode81) [reply tianshou/data/collector.py]: The flag `no_grad` is also unnecessary and should be removed from `Collector.collect`.

Comment (opcode81) [reply tianshou/data/collector.py]: > Actually, Max and I have now concluded that there is really no case where the collector should ever use the policy in train mode. The only reason why it currently needs to be in train mode is that some policies have a deterministic_eval flag, which is True by default (e.g. SAC) and which should basically never be active when collecting data for training.

Max and I have resolved this. See recent commits.

File: tianshou/trainer/utils.py [hunk 9]
```diff
@@ -47,8 +44,8 @@ def gather_info(
     gradient_step: int,
     best_reward: float,
     best_reward_std: float,
-    train_collector: Collector | None = None,
-    test_collector: Collector | None = None,
+    train_collector: BaseCollector | None = None,
+    test_collector: BaseCollector | None = None,
 ) -> InfoStats:
     """A simple wrapper of gathering information from collectors.
```
Comment (MischaPanch) [inline tianshou/trainer/utils.py:51]: note to self: remove docstring that documents info-stats members

File: tianshou/policy/base.py [hunk 10]
```diff
@@ -511,7 +530,8 @@ def update(
         batch, indices = buffer.sample(sample_size)
         self.updating = True
         batch = self.process_fn(batch, buffer, indices)
-        training_stat = self.learn(batch, **kwargs)
+        with in_train_mode(self):
```
Comment (MischaPanch) [inline tianshou/policy/base.py]: shouldn't update or rather learn temporary set the `is_within_training_step` to True?

@opcode81 @maxhuettenrauch

Comment (MischaPanch) [reply tianshou/policy/base.py]: Ah, now I see, it's the correct behavior. We should still adjust it to ensure that it's set to False -with a private contextmanager?

In the current implementation in PGPolicy and so, it has the logic of `within_training_rollout`, where we want to sample from the dist. Here we want to compute gradients, so we are no longer within a rollout.

We are still within a training iteration, so I feel like the name is confusing. How about `within_training_rollout` or `within_training_data_collection`?

File: tianshou/trainer/base.py [hunk 11]
```diff
@@ -426,64 +408,98 @@ def test_step(self) -> tuple[CollectStats, bool]:
 
         return test_stat, stop_fn_flag
 
-    def train_step(self) -> tuple[CollectStats, bool]:
-        """Perform one training step.
+    @contextmanager
+    def _is_training_step_enabled(self, is_training_step: bool):
```
Comment (MischaPanch) [inline tianshou/trainer/base.py]: see my comment above - rename to `_in_training_rollout`?

File: tianshou/policy/base.py (hunk 10 above)
Comment (opcode81) [reply tianshou/policy/base.py]: No, `is_within_training_step` has precisely the semantics we want:

image

The term **training step** should become part of a _glossary_ of well-defined terms in Tianshou.
It's already used in, for example, SamplingConfig - and now it is also reflected in the Trainer (method `training_step`).

> We should still adjust it to ensure that it's set to False -with a private contextmanager?

This is very unclear. Adjust what, where, and why?

File: tianshou/trainer/base.py (hunk 11 above)
Comment (opcode81) [reply tianshou/trainer/base.py]: See comment above.

File: tianshou/policy/base.py (hunk 10 above)
Comment (MischaPanch) [reply tianshou/policy/base.py]: After discussion: the resolution is to raise an error if the flag is false when update is called

File: tianshou/data/collector.py (hunk 8 above)
Comment (MischaPanch) [reply tianshou/data/collector.py]: I will remove no_grad as part of this PR

File: tianshou/trainer/utils.py (hunk 9 above)
Comment (MischaPanch) [reply tianshou/trainer/utils.py:51]: Done

File: tianshou/policy/base.py (hunk 10 above)
Comment (MischaPanch) [reply tianshou/policy/base.py]: Done

File: tianshou/data/collector.py (hunk 7 above)
Comment (MischaPanch) [reply tianshou/data/collector.py]: Done

File: tianshou/data/collector.py (hunk 5 above)
Comment (MischaPanch) [reply tianshou/data/collector.py:239]: Done

File: tianshou/data/collector.py (hunk 6 above)
Comment (MischaPanch) [reply tianshou/data/collector.py:385]: Done

**Review by MischaPanch — APPROVED** [review]

**Comment by MischaPanch** [comment]
This is done, will merge without squashing when tests have finished
